pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+                self.output_conversion = np.squeeze
+            else:
+                self.output_conversion = float
+            self.output_dim = 1
+            res = (res,)
+        elif isinstance(res, Sequence) and res and all(isinstance(x, plo.Observable) for x in res):
+            # for multiple observables values, any valid Python sequence of observables
+            # (i.e., lists, tuples, etc) are supported in the QNode return statement.
+
+            # Device already returns the correct numpy array, so no further conversion is required
+            self.output_conversion = np.asarray
+            self.output_dim = len(res)
+            res = tuple(res)
+        else:
+            raise QuantumFunctionError("A quantum function must return either a single measured observable "
+                                       "or a nonempty sequence of measured observables.")
+
+        # check that all returned observables have a return_type specified
+        for x in res:
+            if x.return_type is None:
+                raise QuantumFunctionError(
+                    "Observable '{}' does not have the measurement type specified.".format(x))
+
+        # check that all ev's are returned, in the correct order
+        if res != tuple(self.obs_queue):
+            raise QuantumFunctionError(
+                "All measured observables must be returned in the order they are measured.")
+
+        # check that no wires are measured more than once
+        m_wires = list(w for ob in res for w in ob.wires)
+        if len(m_wires) != len(set(m_wires)):
+            raise QuantumFunctionError('Each wire in the quantum circuit can only be measured once.')
+
+        self.ops = self.queue + list(res)
+        del self.queue
+        del self.obs_queue
+
+        # True if op is a CV, False if it is a discrete variable (Identity could be either)
+        are_cvs = [isinstance(op, plo.CV) for op in self.ops if not isinstance(op, plops.Identity)]
+        if not all(are_cvs) and any(are_cvs):
+            raise QuantumFunctionError(
+                "Continuous and discrete operations are not allowed in the same quantum circuit.")
+
+        # TODO: we should enforce plugins using the Device.capabilities dictionary to specify
+        # whether they are qubit or CV devices, and remove this logic here.
+        #self.type = 'cv' if all(are_cvs) else 'qubit'  #: str: circuit type, in {'cv', 'qubit'}
+
+
+    def _sort_args(self, args, kwargs):
+        """NEW: Sort the quantum function arguments to positional and keyword-only.
+
+        When the QNode is evaluated via :meth:`~QNode.__call__`, ``kwargs`` may also contain
+        differentiable POSITIONAL_OR_KEYWORD arguments given using the keyword syntax,
+        in addition to KEYWORD_ONLY arguments.
+
+        This method moves all positional arguments from ``kwargs`` to ``args``,
+        and substitutes missing positional arguments with special singleton markers.
```

Comment:
I'm trying to work out the consequence of this change... kwargs can now be differentiable, and `**kwargs` are _required_ to pass non-differentiable data? If so, this is a big change

Reply:
No, the formal arguments `*args` and `**kwargs` in the signature of `QNode.__call__` only relate to how the user decided to input the arguments (positional or keyword syntax), not what those arguments are in the signature of the quantum function. This disparity is why I proposed the new syntax.

So, in the old system auxiliary (non-differentiable) parameters are exactly those parameters that have default values in qfunc signature, and are supposed to be passed using the keyword syntax. The differentiable parameters have to be positional.

In the new system, differentiable parameters are still required to be positional, but now auxiliary arguments are the same as the KEYWORD_ONLY parameters (and possibly also VAR_KEYWORD) in the qfunc signature. Now any argument can have a default value, and auxiliary args must still be passed using the keyword syntax. This method only exists to handle the case where the user decided to pass a POSITIONAL_OR_KEYWORD arg using keyword syntax for some reason.

Reply:
>  and auxiliary args must be passed using the keyword syntax. 

So, the `,*,` suggested change? Like as follows?

```python
def f(diff1, diff2, diff3=x3, diff4=x4, *, non_diff1=y1,...):
```

Reply:
Yes, exactly.